lib/code_corps/accounts/user_invites.ex

```diff
@@ -0,0 +1,100 @@
+defmodule CodeCorps.Accounts.UserInvites do
+  alias CodeCorps.{Project, ProjectUser, Repo, User, UserInvite}
+  alias Ecto.{Changeset, Multi}
+
+  import Ecto.Query
+
+  @spec create_invite(map) :: {:ok, UserInvite.t} | {:error, Changeset.t}
+  def create_invite(%{} = params) do
+    %UserInvite{}
+    |> Changeset.cast(params, [:email, :name, :role, :inviter_id, :project_id])
+    |> Changeset.validate_required([:email, :inviter_id])
+    |> Changeset.validate_inclusion(:role, ~w(contributor admin owner))
+    |> Changeset.assoc_constraint(:inviter)
+    |> Changeset.assoc_constraint(:project)
+    |> ensure_email_not_owned_by_member()
+    |> Repo.insert
+  end
+
+  @spec ensure_email_not_owned_by_member(Changeset.t) :: Changeset.t
+  defp ensure_email_not_owned_by_member(%Changeset{} = changeset) do
+    email = changeset |> Changeset.get_change(:email)
+    project_id = changeset |> Changeset.get_change(:project_id)
+
+    case [email, project_id] do
+      [nil, _] -> changeset
+      [_, nil] -> changeset
+      [email, project_id] ->
+        count =
+          ProjectUser
+          |> where(project_id: ^project_id)
+          |> join(:inner, [pu], u in User, pu.user_id == u.id)
+          |> where([_pu, u], u.email == ^email)
+          |> select([pu, _U], count(pu.id))
+          |> Repo.one
+
+        if count > 0 do
+          changeset |> Changeset.add_error(:email, "Already associated with a project member")
+        else
+          changeset
+        end
+    end
+  end
+
+  @spec claim_invite(map) :: {:ok, User.t}
+  def claim_invite(%{} = params) do
```

Comment:
We'll need to accommodate invites for projects and general invites to Code Corps.

Reply:
Right now, the code sort of does that. It will

- create a user
- join that user with the project, provided the project and role are specified
- assign that user to the assignee

I'm guessing an email will also be involved, which should be pluggable into this process